Document wildcard for network interfaces in Network Settings

GoodMirek opened the pull request:
This issue was supposed to be fixed already, as mentioned in already closed https://github.com/elastic/elasticsearch/issues/14607. This is actually its duplicate.

<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
Yes
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
I believe so
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
NA
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
NA
- If submitting code, have you checked that your submission is for an [OS that we support](https://www.elastic.co/support/matrix#show_os)?
NA
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.
NA

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

javanna commented:
cc @elastic/es-core-infra

javanna commented:
ping @elastic/es-core-infra can we merge this one?

danielmitterdorfer reviewed (changes requested):
Thanks for your PR and I am very sorry that this has fallen through the cracks for so long.

I can see the intent behind your change but I'm not sure whether this is the right place to document this. Yes, `0.0.0.0` is a specialized address according to [RFC 5735](https://tools.ietf.org/html/rfc5735) but in that part of our documentation we talk about values that are treated specially by Elasticsearch itself (outside of any standard). Furthermore, in the following section we refer to these special values again and mention that one can add `:ipv4` and `:ip6` suffixes to special values which is not possible for `0` (or `0.0.0.0`). I think what would work though is to mention that `0.0.0.0` is a possible value in the section `network.host`?

mirek-ppup committed:
Document wildcard for network interfaces in Network Settings

GoodMirek commented:
@danielmitterdorfer I hope I have understood you well. I have rebased and made changes.

I have moved `0.0.0.0` to the section `network.host`, as recommended in the review. I left `0` among special values, as far as I understand it is a special value. I have added a comment related to the applicability of specifiers to this special value, as mentioned in the review.
I have not personally tested or reviewed in the source code that specifiers `:ipv4` and `:ipv6` cannot be used with the special value `0`; I rely upon the review.

danielmitterdorfer commented on docs/reference/modules/network.asciidoc:
```diff
@@ -80,6 +80,9 @@ The following special values may be passed to `network.host`:
 
   Any globally-scoped addresses on the system, for example `8.8.8.8`.
+  
+  `0`::
```
The fact that `0` is recognized at all is due to [IP address normalization in OpenJDK](http://hg.openjdk.java.net/jdk/jdk/file/190b77982361/src/java.base/share/classes/sun/net/util/IPAddressUtil.java#l41) and I don't think that we should document that behavior as a "special value". Special values documented here are meant as "specially treated by Elasticsearch itself", i.e. those have a  [specific handling in the `NetworkService`](https://github.com/elastic/elasticsearch/blob/a9641960e1be101c0bcd4e9bbd28c3a2a2cabeb5/server/src/main/java/org/elasticsearch/common/network/NetworkService.java#L209-L226).

danielmitterdorfer commented on docs/reference/modules/network.asciidoc:
```diff
@@ -24,7 +24,7 @@ host to other nodes in the cluster. Accepts an IP address, hostname, a
 these. Note that any values containing a `:` (e.g., an IPv6 address or
 containing one of the <<network-interface-values,special value>>) must be quoted
-because `:` is a special character in YAML.
+because `:` is a special character in YAML. `0.0.0.0` is an acceptable IP address, it has the same effect as the special value `0`.
```
I think this is fine. As I still would not document `0` as "special value" (see rationale below), I think you could merge the text to say something along the lines of:

> `0.0.0.0` is an acceptable IP address and will binds to all network interfaces.

Can you please also wrap text after 80 characters in docs?

GoodMirek replied:
My primary motivation for this PR is to document the value `0`, as I found that in some config examples and had hard time to find what that means.

Do you agree with:
> `0.0.0.0` is an acceptable IP address and will bind to all network interfaces. Value `0` has the same effect as value `0.0.0.0`.

?

danielmitterdorfer replied:
Yes, I am ok with this change. Just a very minor nit; the articles are missing, i.e.

> The value `0` has the same effect as the value `0.0.0.0`.

danielmitterdorfer replied:
@GoodMirek I think we're very close to completion now. Are you still interested in getting your change in? :)

GoodMirek replied:
@danielmitterdorfer Will look next week. What is a process for merging these changes to current master or whatever the development branch is?
Will I have to go through the whole process again?

danielmitterdorfer replied:
Thanks @GoodMirek. There is no need to concern yourself with other branches than this one. I will take care of merging this to the correct branches on your behalf via cherry-picking (so you will still show up as the author of the change).

danielmitterdorfer replied:
@GoodMirek Any chance you can push the commit including our final changes? Then I can take it from there.

danielmitterdorfer reviewed (changes requested):
Thanks for iterating @GoodMirek. The change makes sense to me. I still think we should not document `0` as special value and included a rationale inline as well as one more suggestion for the docs. Can you please have a look? I think we need another iteration but are close to merging it.

danielmitterdorfer committed:
Merge remote-tracking branch 'origin/6.2' into patch-1

danielmitterdorfer committed:
Add changes as discussed on Github

With this commit we add the necessary changes as discussed in the Github
PR #28839.

danielmitterdorfer committed:
Don't mention `0` as a special value

With this commit we don't mention that `0` is a special value (because
it is not in the sense that Elasticsearch has implemented a special
treatment for `0`).

danielmitterdorfer reviewed (approved):
Thanks for the PR! I went ahead and merged the latest `6.2` branch and also did the changes we mentioned during the review so we're able to merge this. LGTM
